pkg/apis/core/v1alpha1/types_shootstate.go

```diff
@@ -56,11 +56,22 @@ type ShootStateSpec struct {
 	// +optional
 	Gardener []GardenerResourceData `json:"gardener,omitempty" patchStrategy:"merge" patchMergeKey:"name" protobuf:"bytes,1,rep,name=gardener"`
 	// Extensions holds the state of custom resources reconciled by extension controllers in the seed
+	// +patchMergeKey=kind
+	// +patchStrategy=merge
+	// +listType=map
+	// +listMapKey=kind
+	// +listMapKey=name
+	// +listMapKey=purpose
 	// +optional
-	Extensions []ExtensionResourceState `json:"extensions,omitempty" protobuf:"bytes,2,rep,name=extensions"`
+	Extensions []ExtensionResourceState `json:"extensions,omitempty" patchStrategy:"merge" patchMergeKey:"kind" protobuf:"bytes,2,rep,name=extensions"`
```

Comment:
Hm, there can be multiple entries with the same kind. Is this problematic?

Reply:
the following entries above the field should take care of that:
```
// +listType=map
// +listMapKey=kind
// +listMapKey=name
// +listMapKey=purpose
```

I tested it by editing the `status.state` of the controlplane and controlplane exposure extension resources and got both of the entries in the shootstate.

Probably makes sense to add it in the unit tests although I use the fake client there.

Reply:
Hm, can you explain more about these comments/settings? I haven't seen them yet, I think 👀 . Where do their effects manifest?

Reply:
I read a bit more and i've missed a pretty important part about the documentation on those comments/settings. They are only used with "server side apply". So unless I use that in the Patch call there is no reason for them to be in the shootstate types def.

Also the reason I got both entries was that i wasn't using the correct patch strategy as Tim pointed in his [comment](https://github.com/gardener/gardener/pull/3550#discussion_r577363249). 

Anyway after playing some more here's what I found out that there is a problem if I want to generate the patch data for a strategic merge patch with kutil.CreateTwoWayMergePatch(...)  and the extension's state already exists in the `ShootState` and only a part of the new state is changed. As an example if we have the following in the `ShootState (this only happens if the RawExtension holds an array):
```
-kind: worker
  state:
    machineDeployments:
    - entry1
    - entry2
```
And we try to create patch data with from a `ShootState` with modified extension state looking like this:
```
-kind: worker
  state:
    machineDeployments:
    - entry1
    - entry3
```
Generating the patch data will fail with: 
```
Unable to find api field in struct RawExtension for the json field \"machineDeployments\" 
```
There are other subtle errors - if in the example above the new state contains a new struct entry that should replace `machineDeployments` it will instead be added to the state.

The internals of CreateTwoWayMergePatch() is quite complex though and I'm not sure why that's happening yet (would need more time to understand it)

A json-merge patch with optimistic locking seems to do the job, and we won't have to worry about extensions with the same kind and name. @rfranzke @timebertt should we just go for that?

Reply:
> A json-merge patch with optimistic locking seems to do the job, and we won't have to worry about extensions with the same kind and name.

Yeah, probably that's the easiest way to go for now and should already be sufficient, I think.
If it still makes some troubles, we can think about `strategic-merge-patch` or `server-side-apply` again.